query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java

```diff
@@ -48,12 +48,17 @@ boolean updateKnownTypesIfNeeded(final Object value) {
    }
 
    private void handleOnDemandRegistration(Class<?>... classes) {
-      final Class<?>[] toAdd = filterAlreadyIndexed(classes);
-      if (toAdd.length == 0) {
-         return;
+      ArrayList<Class<?>> notIndexedYet = new ArrayList<>(classes.length);
+      for (Class<?> type : classes) {
+         if (!clusterRegistry.containsKey(type)) {
+            notIndexedYet.add(type);
+         }
+      }
+      if (!notIndexedYet.isEmpty()) {
+         Class<?>[] toAdd = notIndexedYet.toArray(new Class[notIndexedYet.size()]);
+         updateSearchFactory(toAdd);
+         updateClusterRegistry(toAdd);
       }
```

Comment:
I think this snippet is crying for some java 8 love :smile:, but I understand if you'd like to keep as is for eventual new backports